src/ray/common/memory_monitor.cc

```diff
@@ -111,41 +111,89 @@ std::tuple<int64_t, int64_t> MemoryMonitor::GetMemoryBytes() {
   return std::tuple(system_used_bytes, system_total_bytes);
 }
 
-int64_t MemoryMonitor::GetCGroupV1MemoryUsedBytes(const char *path) {
+int64_t MemoryMonitor::GetCGroupV1MemoryUsedBytes(const char *stat_path,
+                                                  const char *usage_path) {
+  // Formula to calculate in-used memory from cgroup memory info file:
+  //  used_memory = `memory.usage_in_bytes` - `memory.stat.total_cache - memory.stat.total_shmem`
+  //
+  // This value is consistent with values `MemTotal` `MemAvailable` `MemFree` in `/proc/meminfo`
+  //  and they have relationship of:
+  //  - `memory.usage_in_bytes` == `MemTotal` - `MemFree`
+  //  - `memory.stat.total_cache` == `MemAvailable` - `MemFree`
+  //  - `memory.stat.total_cache` == OS_managed_cache_and_buffer + `shmem`
+  //
+  // Explanation: What's this part `OS_managed_cache_and_buffer` memory for and why
+  //   we should treat this part memory as "Not-in-used" ?
+  // Linux OS tries to fully use the whole physical memory size, so that in many
+  // cases we can observe that the system has very little `MemFree` size,
+  // but at the same time system might have a large `MemAvailable` size,
+  // then this part `MemAvailable - MemFree` size is borrowed by Linux OS
+  // to cache pages / buffers , BUT, once user process requests to allocate memory,
+  // and there is no sufficient free memory, OS will evict cache data out of this
+  // part memory and allocate them to user proces.
+  //
+  // Explanation: What's the part of `shmem` and why we should treat it as "in-used" ?
+  //   `shmem` overview: https://man7.org/linux/man-pages/man7/shm_overview.7.html
+  //   Ray object store use `/dev/shm`, which is a `tmpfs` file system,
+  //   In linux, `tmpfs` file system uses `shmem` memory.
+  //   Note that `tmpfs` file system might swap data to disk (when option noswap=False)
+  //   but `shmem` value always means the `tmpfs` data size in physical memory.
+  //   We can read `shmem` value from /proc/meminfo `Shmem` item or from
+  //   /sys/fs/cgroup/memory/memory.stat `total_shmem` item.
```

Comment:
NOTE: 

Current master Ray code calculation counts "OS_managed_cache_and_buffer" as "in-used" memory, which causes the issue that it heavily underestimate available memory,
and it leads to Ray memory monitor misleadingly triggered to kill Ray tasks while there are lots of available memory.

This is a very big issue.

Because for many IO-heavy tasks, it can quickly fill up physical memory with "OS_managed_cache_and_buffer", then Ray monitor will misleadingly think OS has very few available memory and then kill Ray tasks.